src/visitable.cpp

```diff
@@ -72,17 +73,149 @@ bool visitable<T>::has_item_with( const std::function<bool( const item & )> &fil
             }
         }
         return VisitResponse::NEXT;
     } );
-    return qty == 0;
+    return std::min( qty, limit );
+}
+
+static int has_quality_from_vpart( const vehicle& veh, int part, const std::string& qual, int level, int limit )
+{
+    int qty = 0;
+
+    auto pos = veh.parts[ part ].mount;
+    for( const auto &n : veh.parts_at_relative( pos.x, pos.y ) ) {
+
+        // only unbroken parts can provide tool qualities
+        if( veh.parts[ n ].hp > 0 ) {
+            auto tq = veh.part_info( n ).qualities;
+            auto iter = tq.find( qual );
+
+            // does the part provide this quality?
+            if( iter != tq.end() && iter->second >= level ) {
+                if( ++qty >= limit ) {
+                    break;
+                }
+            }
+        }
+    }
+    return std::min( qty, limit );
+}
+
+template <typename T>
+bool visitable<T>::has_quality( const std::string &qual, int level, int qty ) const
+{
+    return has_quality_internal( *this, qual, level, qty ) == qty;
+}
+
+template <>
+bool visitable<vehicle_selector>::has_quality( const std::string &qual, int level, int qty ) const
+{
+    for( const auto& cursor : static_cast<const vehicle_selector &>( *this ) ) {
+        qty -= has_quality_from_vpart( cursor.veh, cursor.part, qual, level, qty );
+        if( qty <= 0 ) {
+            return true;
+        }
+    }
+    return has_quality_internal( *this, qual, level, qty ) == qty;
+}
+
+template <>
+bool visitable<vehicle_cursor>::has_quality( const std::string &qual, int level, int qty ) const
+{
+    auto self = static_cast<const vehicle_cursor *>( this );
+
+    qty -= has_quality_from_vpart( self->veh, self->part, qual, level, qty );
+    return qty <= 0 ? true : has_quality_internal( *this, qual, level, qty ) == qty;
+}
+
+template <typename T>
+static int max_quality_internal( const T& self, const std::string &qual )
+{
+    int res = 0;
+    self.visit_items( [&res,&qual]( const item *e ) {
+        res = std::max( res, e->get_quality( qual ) );
+        return VisitResponse::NEXT;
+    } );
+    return res;
+}
+
+static int max_quality_from_vpart( const vehicle& veh, int part, const std::string& qual )
+{
+    int res = 0;
+
+    auto pos = veh.parts[ part ].mount;
+    for( const auto &n : veh.parts_at_relative( pos.x, pos.y ) ) {
+
+        // only unbroken parts can provide tool qualities
+        if( veh.parts[ n ].hp > 0 ) {
+            auto tq = veh.part_info( n ).qualities;
+            auto iter = tq.find( qual );
+
+            // does the part provide this quality?
+            if( iter != tq.end() ) {
+                res = std::max( res, iter->second );
+            }
+        }
+    }
+    return res;
+}
+
+template <typename T>
+int visitable<T>::max_quality( const std::string &qual ) const
+{
+    return max_quality_internal( *this, qual );
+}
+
+template<>
+int visitable<Character>::max_quality( const std::string &qual ) const
+{
+    int res = 0;
```

Comment:
The old function in player started with INT_MIN, not 0. This function now yields different behavior, noticeable when butchering: `game::butcher` checks for `BUTCHER quality > INT_MIN` (which indicates a tool with that quality is available). This function now will always return something `>= 0`, and butchering will *always* be possible.

Note that negative butchering qualities are valid (and actually used), it decreases the amount of resulting items and increases the duration of the activity.